Remora.Rest.Generators/FluentJsonContextSourceGenerator.cs

```diff
@@ -4,123 +4,156 @@
-                {
-                    // probably raise a diagnostic?
-                    continue;
-                }
-            }
+        if (invocation.Expression is not MemberAccessExpressionSyntax access)
+        {
+            return false;
+        }
 
-            if (contextTypeSymbol.BaseType?.Name != "JsonSerializerContext")
-            {
-                // diagnostic or error out?
-                continue;
-            }
+        if (access.Name is not GenericNameSyntax generic)
+        {
+            return false;
+        }
 
-            var converterInitializations = CreateConverterInitializations(context, contextTypeSymbol);
+        if (generic.Identifier.Text != "WithContext")
+        {
+            return false;
+        }
 
-            if (converterInitializations.Count <= 0)
-            {
-                // no optionals in use here
-                continue;
-            }
+        if (generic.TypeArgumentList.Arguments.Count != 1)
+        {
+            return false;
+        }
+
+        if (generic.TypeArgumentList.Arguments[0] is not IdentifierNameSyntax ins)
+        {
+            return false;
+        }
+
+        return true;
+    }
+
+    private static JsonTypeModel TransformSyntaxNode(GeneratorSyntaxContext context)
+    {
+        var node = (InvocationExpressionSyntax)context.Node;
+        var access = (MemberAccessExpressionSyntax)node.Expression;
+        var generic = (GenericNameSyntax)access.Name;
+
+        var identifier = (IdentifierNameSyntax)generic.TypeArgumentList.Arguments[0];
 
-            var contextNamespace = GetContextDeclarationNamespace(context, contextTypeSymbol);
-
-            var source = SourceText.From
-            (
-                $$"""
-                  // <auto-generated/>
-                  #pragma warning disable CS0618
-
-                  using System.Text.Json;
-                  using Remora.Rest.Json.Configuration;
-                  using Remora.Rest.Json.Converters;
-                  {{contextNamespace}}
-                  public partial class {{contextType.Type.Name}} : IAdditionalConverterProvider
-                  {
-                      public IReadOnlyList<JsonConverter> IAdditionalConverterProvider.AdditionalConverters { get; } = new JsonConverter[]
-                      {
-                          {{string.Join(",\n        ", converterInitializations)}}
-                      };
-                  }
-                  """,
-                encoding: Encoding.UTF8
-            );
-
-            var hintName = $"{contextType.Type.Name}.AdditionalConverters.g.cs";
-            context.AddSource(hintName, source);
+        var typeInfo = context.SemanticModel.GetTypeInfo(identifier);
+
+        if (typeInfo.Type is not INamedTypeSymbol type)
+        {
+            // diagnostic? there should ideally be a separate analyzer catching all of these
+            return default;
+        }
+
+        if (type.BaseType?.Name != "JsonSerializerContext")
```

Comment:
Could we use `GetTypeByMetadataName` here as well?

Reply:
most likely yes.

Reply:
I think I'd like that - saves us a bit of trouble with duck typing.